testutils/passes: add utility for nolint comment, add support in unconvert, descriptormarshal

ajwerner opened the pull request:
See individual commits. This PR is in support of unblocking https://github.com/cockroachdb/cockroach/pull/54946.

Commit:
lint/passes: refactor common code into passesutil

Functions to determine the containing file and to determine whether
an ast node has a `nolint` comment should not live inside of a single
pass implementation.

There's a bunch of testing of the logic here in the subsequent two
commits.

Release note: None

Commit:
testutils/lint/passes/descriptormarshal: adopt nolint to defeat linter

Prior to this change there was hard-coded logic inside the linter to deal with
these cases. It was gnarly. Now that there are tools to easily detect nolint
comments, use them.

Release note: None

Commit:
testutils/lint/passes/unconvert: add support for nolint comments

Adopts the earlier commit to add support for nolint comments to defeat the
linter.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/55218)
<!-- Reviewable:end -->

Commit:
fixup! lint/passes: refactor common code into passesutil

Commit:
testutils/lint/passes/passesutil: add a test

Invoke the tests from the client packages as a way to get coverage and
exercise the logic.

Release note: None

knz reviewed (approved):
> Reviewed 4 of 4 files at r1, 5 of 5 files at r2, 2 of 2 files at r3, 2 of 2 files at r4, 1 of 1 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55218)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @pbardea)

___
*[pkg/testutils/lint/passes/unconvert/unconvert.go, line 87 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/55218#-MIxp5UaBkxwecc2zHOS:-MIxp5UaBkxwecc2zHOT:bp48w3e) ([raw file](https://github.com/cockroachdb/cockroach/blob/5852fa90614740bfd0fbe54210b46ca7d9b8c987/pkg/testutils/lint/passes/unconvert/unconvert.go#L87)):*
> ```Go
> 			}
> 		}
> 		if passesutil.HasNolintComment(pass, call, name) {
> ```

Can you explain (to me) what is the benefit of this new feature? I don't know what we have any code that wants to skip the convert linter.


<!-- Sent from Reviewable.io -->

ajwerner commented:
See https://github.com/cockroachdb/cockroach/pull/54946.

> An explicit floating-point type conversion rounds to the precision of the target type, preventing fusion that would discard that rounding.

From https://golang.org/ref/spec

ajwerner commented:
TFTR!

bors r=knz
